packages/services/src/view-service/util/parallel.ts

```diff
@@ -0,0 +1,77 @@
+import { JsonObject, JsonValue } from '@bufbuild/protobuf';
+import { FullViewingKey } from '@penumbra-zone/protobuf/penumbra/core/keys/v1/keys_pb';
+import {
+  Action,
+  TransactionPlan,
+  WitnessData,
+} from '@penumbra-zone/protobuf/penumbra/core/transaction/v1/transaction_pb';
+import { WorkerBuildAction } from './build-action-worker.js';
+import { ConnectError } from '@connectrpc/connect';
+
+/**
+ * Build actions in parallel by launching a worker for each action in the plan.
+ * @returns An individually-promised list of build results.
+ */
+export const launchActionWorkers = (
+  transactionPlan: TransactionPlan,
+  witness: WitnessData,
+  fullViewingKey: FullViewingKey,
+  signal: AbortSignal,
+): Promise<Action>[] => {
+  const partialRequest: Omit<WorkerBuildAction, 'actionPlanIndex'> = {
+    transactionPlan: transactionPlan.toJson() as JsonObject,
+    witness: witness.toJson() as JsonObject,
+    fullViewingKey: fullViewingKey.toJson() as JsonObject,
+  };
+
+  const workerOutputs = transactionPlan.actions.map(async (_, actionPlanIndex) => {
+    const buildReq: WorkerBuildAction = {
+      ...partialRequest,
+      actionPlanIndex,
+    };
+
+    const actionWorker = new Worker(new URL('./build-action-worker.js', import.meta.url));
```

Comment:
**comment:** previously we were spawning the worker in the extension's [offscreen handler](https://github.com/prax-wallet/web/pull/190/files#diff-8c2640a27ba27f5d2d0715fe7abd9f49980168a00aae7d28528e5daf1c4c316bL48), and now it's happening directly in the services package. 

Based on offline discussion, a generic offscreen utility called [OffscreenWorker](https://github.com/prax-wallet/web/pull/190/files#diff-e510b8ebc1f7156cc731b9eeea856aaf3081a4d5e82c0f38a2a6afbb751dc408R5) now manages the lifecycle of the offscreen document. A polyfill was created where `new Worker` references the global `Worker` variable. So when the service worker sets [globalThis.Worker = OffscreenWorker](https://github.com/prax-wallet/web/pull/190/files#diff-b9be68cac5d37d21289e20d8d7fcb88d993de52b16e69a6500255f3fde092ae0R48), all `Worker` references in the script now point to `OffscreenWorker`. Each new `Worker` call creates an instance of `OffscreenWorker`, which communicates via a dedicated port to an actual `Worker` running in the offscreen document.

Reply:
we should also benchmark the web worker performance to identify any potential regressions in the proving time, though there shouldn't be any.